### PR TITLE
Zonal averages, horizontal averages, and volume averages under the same structure

### LINKED ISSUE #840: `versioninfo_with_gpu` util fails on machine with CUDA but running CPU models
We should only add GPU data is there's a current CUDA context, i.e.

```julia
if !isnothing(CUDA.CuCurrentContext())
    gpu_name = CUDA.CuCurrentContext() |> CUDA.device |> CUDA.name
    s = s * "  GPU: $gpu_name\n"
end
```

### PR DESCRIPTION
**Pull request by sandreza**
Adds abstract averages: zonal averages, horizontal averages, and volume averages under the same structure

Resolves #840

### CODE REVIEW

File: src/AbstractOperations/computations.jl [hunk 1]
```diff
@@ -91,11 +121,38 @@ Returns the representation of a `HorizontalAverage` over the operation `op`, usi
 HorizontalAverage(op::AbstractOperation, model::AbstractModel; kwargs...) =
     HorizontalAverage(op, model.pressures.pHY′; kwargs...)
 
-"""Compute the horizontal average of a computation."""
+
+"""
+    ZonalAverage(op::AbstractOperation, model; kwargs...)
+Returns the representation of a `ZonalAverage` over the operation `op`, using
```
Comment (glwagner) [inline src/AbstractOperations/computations.jl:127]: White space missing here

**Comment by glwagner** [comment]
We need tests for both the zonal average and the volume average before this can be merged.

Side note: we probably want to do away with the notation "Zonal" and "Horizontal". It might be better to use something like `XAverage` and `XYAverage`. This naturally extends to the other possible averages, `YAverage`, `YZAverage`, etc)

**Comment by glwagner** [comment]
I changed the name of the PR to better reflect its changes. Does this PR replace #783 ?

**Comment by ali-ramadhan** [comment]
I think I'd be on board with `XAverage`, `XYAverage`, etc. as Oceananigans.jl is not on the sphere so there's no notion of a zonal or meridional direction (unless maybe when implied by an `FPlane` or `BetaPlane`).

I do prefer `HorizontalAverage` over `ZAverage` though, but would rather not have a lot of aliases floating around...

> We need tests for both the zonal average and the volume average before this can be merged.

Yeah for sure. @sandreza and I were gonna add those next time we pair program. Also reduce the amount of boilerplate as you suggested in your review of #783.

> Does this PR replace #783?

Yes so I closed #783 and left an explanation.

**Comment by glwagner** [comment]
> I do prefer HorizontalAverage over ZAverage though, but would rather not have a lot of aliases floating around...

I think with this nomenclature, what we now call `HorizontalAverage` would become `XYAverage`.

Another possibility is to simply have `Average`, and then introduce a `dim` type parameter that can be `:x`, `:y`, `:xy`, `:xyz`, etc... ? So we'd use

```julia
horizontally_averaged_u = Average{:xy}(model.velocities.u)

zonally_averaged_u = Average{:x}(model.velocities.u)
```

Even better would be to use an abstraction with `dims=(1,)` perhaps, but looking at the code I'm not 100% sure how to do that.

**Comment by ali-ramadhan** [comment]
We can continue with the current nomenclature to avoid breaking changes and discuss again for a future PR.

@glwagner This PR is now ready for review.

We should also run GPU tests manually. The PR was opened from a fork so GitLab CI won't run on this PR.

File: src/OutputWriters/checkpointer.jl [hunk 2]
```diff
@@ -63,7 +63,7 @@ function Checkpointer(model; frequency=nothing, interval=nothing, dir=".",
         p isa Symbol || @error "Property $p to be checkpointed must be a Symbol."
         p ∉ propertynames(model) && @error "Cannot checkpoint $p, it is not a model property!"
 
-        if has_reference(Function, getproperty(model, p))
+        if has_reference(Function, getproperty(model, p)) && (p ∉ required_properties)
```
Comment (glwagner) [inline src/OutputWriters/checkpointer.jl:66]: Is this meant to be here?

File: src/OutputWriters/netcdf_output_writer.jl [hunk 3]
```diff
@@ -94,7 +94,7 @@ function NetCDFOutputWriter(model, outputs; filename,
 
     # Opens the same output file for writing fields from the user-supplied variable outputs
     dataset = Dataset(filename, "a")
-
+    println("hello")
```
Comment (glwagner) [inline src/OutputWriters/netcdf_output_writer.jl:97]: This should be removed?

Comment (ali-ramadhan) [reply src/OutputWriters/netcdf_output_writer.jl:179]: Done.

File: src/OutputWriters/checkpointer.jl (hunk 2 above)
Comment (ali-ramadhan) [reply src/OutputWriters/checkpointer.jl]: Should be gone now. This change should be part of PR #839.

**Comment by ali-ramadhan** [comment]
Resolves #840 now

**Comment by ali-ramadhan** [comment]
I've refactored this PR to combine all the different average diagnostics into one `Average` diagnostic. Different averages are specified with the `dims` keyword argument, e.g. like `Statistics.mean`.

`HorizontalAverage` is now `Average(..., dims=(1, 2))` and `ZonalAverage` is now `Average(..., dims=1)` or `Average(..., dims=(1,))` so we have all possible averages available to us now.

File: src/Diagnostics/average.jl [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+run this diagnostic if it is part of `simulation.diagnostics`. `iteration_interval` is a
+number of iterations while `time_interval` is a time interval in units of `model.clock.time`.
+
+A `return_type` can be used to specify the type returned when the `Average` is
+used as a callable object. The default `return_type=Array` is useful when running a GPU
+model and you want to save the output to disk by passing it to an output writer.
+"""
+function Average(field; dims, iteration_interval=nothing, time_interval=nothing, return_type=Array)
+    dims isa Union{Int, Tuple} || @error "Average dims must be an integer or tuple!"
+    dims isa Int && (dims = tuple(dims))
```
Comment (glwagner) [inline src/Diagnostics/average.jl:46]: Now that's a feature.

Comment (ali-ramadhan) [reply src/Diagnostics/average.jl:46]: Haha `Statistics.mean` does it so I figured we should be fancy too.

**Pull request merged**